File: app/lib/services/score_rule_engine.dart
import '../models/match_models.dart';

class ScoreSnapshot {
  const ScoreSnapshot({
    required this.myGames,
    required this.opponentGames,
    required this.myPoints,
    required this.opponentPoints,
    required this.isFinalGame,
    required this.isCompleted,
    required this.servingSide,
    required this.serverId,
    required this.receiverId,
    required this.shouldChangeSides,
    required this.shouldChangeService,
    required this.gameWinners,
  });
  final int myGames;
  final int opponentGames;
  final int myPoints;
  final int opponentPoints;
  final bool isFinalGame;
  final bool isCompleted;
  final Side servingSide;
  final String serverId;
  final String receiverId;
  final bool shouldChangeSides;
  final bool shouldChangeService;
  final List<Side> gameWinners;
}

/// Flutterに依存しない得点・サービス順の状態遷移。
class ScoreRuleEngine {
  const ScoreRuleEngine();

  ScoreSnapshot evaluate(MatchRecord record) {
    var myGames = 0;
    var opponentGames = 0;
    var myPoints = 0;
    var opponentPoints = 0;
    var pointsInGame = 0;
    var completed = false;
    var changedSides = false;
    var changedService = false;
    final gameWinners = <Side>[];

    for (final event in record.events) {
      if (completed) break;
      changedSides = false;
      changedService = false;
      pointsInGame++;
      if (event.winningSide == Side.mine) {
        myPoints++;
      } else {
        opponentPoints++;
      }
      final finalGame = _isFinal(record.format, myGames, opponentGames);
      if (_winsGame(record.format, myPoints, opponentPoints, finalGame)) {
        final winner = myPoints > opponentPoints ? Side.mine : Side.opponent;
        if (winner == Side.mine) {
          myGames++;
        } else {
          opponentGames++;
        }
        gameWinners.add(winner);
        changedSides = !finalGame && (myGames + opponentGames).isOdd;
        changedService = true;
        completed =
            myGames == record.format.gamesToWin ||
            opponentGames == record.format.gamesToWin;
        if (!completed) {
          myPoints = 0;
          opponentPoints = 0;
          pointsInGame = 0;
        }
      }
      if (finalGame && !completed) {
        changedService = pointsInGame.isEven;
        changedSides =
            pointsInGame == 2 ||
            (pointsInGame > 2 && (pointsInGame - 2) % 4 == 0);
      }
    }

    final finalGame = _isFinal(record.format, myGames, opponentGames);
    final service = _service(
      record,
      myGames + opponentGames,
      pointsInGame,
      finalGame,
    );
    return ScoreSnapshot(
      myGames: myGames,
      opponentGames: opponentGames,
      myPoints: myPoints,
      opponentPoints: opponentPoints,
      isFinalGame: finalGame,
      isCompleted: completed,
      servingSide: service.$1,
      serverId: service.$2,
      receiverId: service.$3,
      shouldChangeSides: changedSides,
      shouldChangeService: changedService,
      gameWinners: List.unmodifiable(gameWinners),
    );
  }

  bool _isFinal(MatchFormatPreset format, int mine, int opponent) =>
      mine == format.maximumGames ~/ 2 && opponent == format.maximumGames ~/ 2;

  bool _winsGame(
    MatchFormatPreset format,
    int mine,
    int opponent,
    bool finalGame,
  ) {
    final target = finalGame ? 7 : 4;
    final noAd = format.isNoAd;
    final high = mine > opponent ? mine : opponent;
    return high >= target && (noAd || (mine - opponent).abs() >= 2);
  }

  (Side, String, String) _service(
    MatchRecord record,
    int completedGames,
    int pointIndex,
    bool finalGame,
  ) {
    final serviceBlock = pointIndex ~/ 2;
    final servingSide = finalGame
        ? serviceBlock.isEven
              ? record.firstServingSide
              : record.firstServingSide.other
        : completedGames.isEven
        ? record.firstServingSide
        : record.firstServingSide.other;
    final servingPair = servingSide == Side.mine
        ? record.myPair
        : record.opponentPair;
    final receivingPair = servingSide == Side.mine
        ? record.opponentPair
        : record.myPair;
    final initialServer = servingSide == record.firstServingSide
        ? record.firstServerId
        : record.firstReceiverId;
    final initialReceiver = servingSide == record.firstServingSide
        ? record.firstReceiverId
        : record.firstServerId;
    final serverRotation = finalGame ? serviceBlock ~/ 2 : pointIndex ~/ 2;
    final server = _alternatePlayer(servingPair, initialServer, serverRotation);
    final receiver = _alternatePlayer(
      receivingPair,
      initialReceiver,
      pointIndex,
    );
    return (servingSide, server, receiver);
  }

  String _alternatePlayer(Pair pair, String preferredId, int index) {
    final preferredIndex = pair.players.indexWhere(
      (player) => player.id == preferredId,
    );
    final base = preferredIndex < 0 ? 0 : preferredIndex;
    return pair.players[(base + index) % pair.players.length].id;
  }
}

File: app/ios/WatchShared/ScoreRuleEngine.swift
import Foundation

struct ScoreSnapshotDTO: Equatable {
  let myGames: Int
  let opponentGames: Int
  let myPoints: Int
  let opponentPoints: Int
  let isFinalGame: Bool
  let isCompleted: Bool
  let servingSide: MatchSide
  let serverId: String
  let receiverId: String
  let shouldChangeSides: Bool
  let shouldChangeService: Bool
}

struct PointContextDTO {
  let servingSide: MatchSide
  let winningSide: MatchSide
  let serveAttempt: ServeAttemptDTO
}

struct SwiftScoreRuleEngine {
  func context(for eventID: String, in record: MatchRecordDTO) -> PointContextDTO? {
    guard let index = record.events.firstIndex(where: { $0.id == eventID }) else {
      return nil
    }
    var previous = record
    previous.events = Array(record.events.prefix(index))
    return PointContextDTO(
      servingSide: evaluate(previous).servingSide,
      winningSide: record.events[index].winningSide,
      serveAttempt: record.events[index].serveAttempt
    )
  }

  func evaluate(_ record: MatchRecordDTO) -> ScoreSnapshotDTO {
    var myGames = 0
    var opponentGames = 0
    var myPoints = 0
    var opponentPoints = 0
    var pointsInGame = 0
    var completed = false
    var changedSides = false
    var changedService = false

    for event in record.events where !completed {
      changedSides = false
      changedService = false
      pointsInGame += 1
      if event.winningSide == .mine { myPoints += 1 } else { opponentPoints += 1 }
      let finalGame = isFinal(record.format, myGames, opponentGames)
      if winsGame(record.deuceEnabled, myPoints, opponentPoints, finalGame) {
        if myPoints > opponentPoints { myGames += 1 } else { opponentGames += 1 }
        changedSides = !finalGame && (myGames + opponentGames).isMultiple(of: 2) == false
        changedService = true
        completed = myGames == record.format.gamesToWin || opponentGames == record.format.gamesToWin
        if !completed {
          myPoints = 0
          opponentPoints = 0
          pointsInGame = 0
        }
      }
      if finalGame && !completed {
        changedService = pointsInGame.isMultiple(of: 2)
        changedSides = pointsInGame == 2 || (pointsInGame > 2 && (pointsInGame - 2).isMultiple(of: 4))
      }
    }

    let finalGame = isFinal(record.format, myGames, opponentGames)
    let service = service(record, myGames + opponentGames, pointsInGame, finalGame)
    return ScoreSnapshotDTO(
      myGames: myGames,
      opponentGames: opponentGames,
      myPoints: myPoints,
      opponentPoints: opponentPoints,
      isFinalGame: finalGame,
      isCompleted: completed,
      servingSide: service.0,
      serverId: service.1,
      receiverId: service.2,
      shouldChangeSides: changedSides,
      shouldChangeService: changedService
    )
  }

  private func isFinal(_ format: MatchFormatDTO, _ mine: Int, _ opponent: Int) -> Bool {
    mine == format.maximumGames / 2 && opponent == format.maximumGames / 2
  }

  private func winsGame(
    _ deuceEnabled: Bool,
    _ mine: Int,
    _ opponent: Int,
    _ finalGame: Bool
  ) -> Bool {
    let target = finalGame ? 7 : 4
    let high = max(mine, opponent)
    return high >= target && (!deuceEnabled || abs(mine - opponent) >= 2)
  }

  private func service(
    _ record: MatchRecordDTO,
    _ completedGames: Int,
    _ pointIndex: Int,
    _ finalGame: Bool
  ) -> (MatchSide, String, String) {
    let serviceBlock = pointIndex / 2
    let servingSide: MatchSide
    if finalGame {
      servingSide = serviceBlock.isMultiple(of: 2) ? record.firstServingSide : record.firstServingSide.other
    } else {
      servingSide = completedGames.isMultiple(of: 2) ? record.firstServingSide : record.firstServingSide.other
    }
    let servingPair = servingSide == .mine ? record.myPair : record.opponentPair
    let receivingPair = servingSide == .mine ? record.opponentPair : record.myPair
    let initialServer = servingSide == record.firstServingSide ? record.firstServerId : record.firstReceiverId
    let initialReceiver = servingSide == record.firstServingSide ? record.firstReceiverId : record.firstServerId
    let serverRotation = finalGame ? serviceBlock / 2 : pointIndex / 2
    return (
      servingSide,
      alternate(servingPair, initialServer, serverRotation),
      alternate(receivingPair, initialReceiver, pointIndex)
    )
  }

  private func alternate(_ pair: PairDTO, _ preferredID: String, _ index: Int) -> String {
    let base = pair.players.firstIndex(where: { $0.id == preferredID }) ?? 0
    return pair.players[(base + index) % pair.players.count].id
  }
}

struct SwiftPointReasonPolicy {
  func availableReasons(
    servingSide: MatchSide,
    winningSide: MatchSide,
    serveAttempt: ServeAttemptDTO
  ) -> [PointReasonDTO] {
    var reasons: [PointReasonDTO] = [.rallyWinner, .opponentNet, .opponentOut, .other]
    if winningSide == servingSide {
      reasons.insert(.serviceAce, at: 0)
    } else {
      reasons.insert(.returnAce, at: 0)
      if serveAttempt == .second { reasons.insert(.opponentDoubleFault, at: 1) }
    }
    return reasons
  }
}
